src/services/permissions-checker.js

```diff
@@ -5,18 +5,63 @@ const VError = require('verror');
 
 let permissionsPerRendering = {};
 
-function PermissionsChecker(environmentSecret, renderingId, collectionName, permissionName) {
+function PermissionsChecker(
+  environmentSecret,
+  renderingId,
+  collectionName,
+  permissionName,
+  smartActionId = null,
+  endpoint = null,
+) {
   const EXPIRATION_IN_SECONDS = process.env.FOREST_PERMISSIONS_EXPIRATION_IN_SECONDS || 3600;
 
-  function isAllowed() {
+  function isCollectionAllowed() {
     const permissions =
       permissionsPerRendering[renderingId] && permissionsPerRendering[renderingId].data;
 
     if (!permissions || !permissions[collectionName] || !permissions[collectionName].collection) {
+      return {
+        allowed: false,
+        error: `'${permissionName}' access forbidden on collection '${collectionName}'`,
+      };
+    }
+
+    if (permissionName === 'execute') {
+      const isAllowed = isSmartActionAllowed();
+      return isAllowed
+        ? { allowed: true }
+        : {
+          allowed: false,
+          error: `Smart action '${getActionName()}' execution forbidden on collection '${collectionName}'`,
+        };
+    }
+
+    const isAllowed = permissions[collectionName].collection[permissionName];
+    if (!isAllowed) {
+      return {
+        allowed: false,
+        error: `'${permissionName}' access forbidden on collection '${collectionName}'`,
+      };
+    }
+
+    return { allowed: true };
+  }
+
+  function isSmartActionAllowed() {
+    const permissions = permissionsPerRendering[renderingId].data;
+
+    if (!permissions[collectionName].smartActions
+      || !permissions[collectionName].smartActions[smartActionId]
+      || !permissions[collectionName].smartActions[smartActionId].endpoint) {
+      return false;
+    }
+
+    if (permissions[collectionName].smartActions[smartActionId].endpoint !== endpoint) {
```

Comment:
We should check the METHOD, you can have two relationships on same endpoint but different method